File: apps/web/src/lib/webdav.ts
/**
 * Minimal WebDAV client covering the verbs the controller needs to read the
 * shared object store: PROPFIND to list, GET to read, PUT/MKCOL to write during
 * connection tests. Agents perform the bulk of writes themselves.
 */

export interface WebdavConfig {
  baseUrl: string;
  username: string;
  password: string;
}

export interface WebdavEntry {
  href: string;
  name: string;
  isDirectory: boolean;
  size: number;
  lastModified: number | null;
}

const PROPFIND_BODY =
  '<?xml version="1.0" encoding="utf-8"?>' +
  '<d:propfind xmlns:d="DAV:"><d:prop>' +
  "<d:resourcetype/><d:getcontentlength/><d:getlastmodified/>" +
  "</d:prop></d:propfind>";

export class WebdavError extends Error {
  constructor(
    message: string,
    readonly status: number | null = null
  ) {
    super(message);
  }
}

/** Normalizes to exactly one trailing slash so joins never double up. */
export function normalizeBaseUrl(raw: string): string {
  const trimmed = raw.trim();
  if (trimmed.length === 0) {
    throw new WebdavError("WebDAV base URL must not be empty");
  }
  let parsed: URL;
  try {
    parsed = new URL(trimmed);
  } catch {
    throw new WebdavError(`WebDAV base URL is not a valid URL: ${trimmed}`);
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new WebdavError("WebDAV base URL must use http or https");
  }
  parsed.hash = "";
  parsed.search = "";
  if (!parsed.pathname.endsWith("/")) {
    parsed.pathname = `${parsed.pathname}/`;
  }
  return parsed.toString();
}

function encodeSegments(path: string): string {
  return path
    .split("/")
    .filter((segment) => segment.length > 0)
    .map((segment) => encodeURIComponent(segment))
    .join("/");
}

export class WebdavClient {
  private readonly baseUrl: string;
  private readonly authorization: string;

  constructor(
    config: WebdavConfig,
    private readonly fetchImpl: typeof fetch = fetch
  ) {
    this.baseUrl = normalizeBaseUrl(config.baseUrl);
    const credential = Buffer.from(
      `${config.username}:${config.password}`,
      "utf8"
    ).toString("base64");
    this.authorization = `Basic ${credential}`;
  }

  url(path: string): string {
    return new URL(encodeSegments(path), this.baseUrl).toString();
  }

  private async request(
    method: string,
    path: string,
    init: { body?: BodyInit; headers?: Record<string, string> } = {}
  ): Promise<Response> {
    let response: Response;
    try {
      response = await this.fetchImpl(this.url(path), {
        method,
        headers: { authorization: this.authorization, ...(init.headers ?? {}) },
        body: init.body,
        redirect: "follow"
      });
    } catch (cause) {
      throw new WebdavError(
        `WebDAV ${method} failed: ${cause instanceof Error ? cause.message : String(cause)}`
      );
    }
    if (response.status === 401 || response.status === 403) {
      throw new WebdavError("WebDAV rejected the credential", response.status);
    }
    return response;
  }

  /**
   * True when the resource exists, tested with a depth-0 PROPFIND.
   *
   * HEAD would be cheaper, but WsgiDAV answers a 404 HEAD with a
   * `Content-Length` for an error page it then correctly does not send, which
   * leaves a keep-alive connection expecting bytes that never arrive; the next
   * response on that connection is unparsable. PROPFIND always sends the body
   * it declares. relaydot/webdav.py avoids HEAD for the same reason.
   */
  async exists(path: string): Promise<boolean> {
    const response = await this.request("PROPFIND", path, {
      headers: { depth: "0", "content-type": "application/xml" },
      body: PROPFIND_BODY
    });
    if (response.status === 404) {
      return false;
    }
    if (response.status !== 207 && !response.ok) {
      throw new WebdavError(
        `WebDAV PROPFIND ${path} returned ${response.status}`,
        response.status
      );
    }
    return true;
  }

  async get(path: string): Promise<Uint8Array | null> {
    const response = await this.request("GET", path);
    if (response.status === 404) {
      return null;
    }
    if (!response.ok) {
      throw new WebdavError(
        `WebDAV GET ${path} returned ${response.status}`,
        response.status
      );
    }
    return new Uint8Array(await response.arrayBuffer());
  }

  async getText(path: string): Promise<string | null> {
    const bytes = await this.get(path);
    return bytes === null ? null : Buffer.from(bytes).toString("utf8");
  }

  async put(path: string, body: Uint8Array | string): Promise<void> {
    const response = await this.request("PUT", path, {
      body: typeof body === "string" ? body : new Uint8Array(body),
      headers: { "content-type": "application/octet-stream" }
    });
    if (!response.ok && response.status !== 204) {
      throw new WebdavError(
        `WebDAV PUT ${path} returned ${response.status}`,
        response.status
      );
    }
  }

  async mkcol(path: string): Promise<void> {
    const response = await this.request("MKCOL", path);
    // 405 means the collection already exists, which is not an error here.
    if (!response.ok && response.status !== 405) {
      throw new WebdavError(
        `WebDAV MKCOL ${path} returned ${response.status}`,
        response.status
      );
    }
  }

  /** Creates every missing ancestor collection of a directory path. */
  async ensureCollection(path: string): Promise<void> {
    const segments = path.split("/").filter((segment) => segment.length > 0);
    let current = "";
    for (const segment of segments) {
      current = current.length === 0 ? segment : `${current}/${segment}`;
      await this.mkcol(current);
    }
  }

  async list(path: string, depth: "1" | "infinity" = "1"): Promise<WebdavEntry[]> {
    const response = await this.request("PROPFIND", path, {
      headers: { depth, "content-type": "application/xml" },
      body: PROPFIND_BODY
    });
    if (response.status === 404) {
      return [];
    }
    if (response.status !== 207 && !response.ok) {
      throw new WebdavError(
        `WebDAV PROPFIND ${path} returned ${response.status}`,
        response.status
      );
    }
    return parsePropfind(await response.text(), this.baseUrl);
  }
}

/**
 * Extracts responses from a multistatus body without a full XML parser. The
 * controller only needs href, collection flag, length, and mtime, and WebDAV
 * multistatus is regular enough that a scan is reliable for those fields.
 */
export function parsePropfind(xml: string, baseUrl: string): WebdavEntry[] {
  const entries: WebdavEntry[] = [];
  const basePath = new URL(baseUrl).pathname;
  const blocks = xml.match(/<[a-zA-Z0-9]*:?response[\s>][\s\S]*?<\/[a-zA-Z0-9]*:?response>/g);
  for (const block of blocks ?? []) {
    const href = tag(block, "href");
    if (href === null) {
      continue;
    }
    let pathname: string;
    try {
      pathname = new URL(href, baseUrl).pathname;
    } catch {
      continue;
    }
    const relative = decodeURIComponent(
      pathname.startsWith(basePath) ? pathname.slice(basePath.length) : pathname
    ).replace(/^\/+|\/+$/g, "");
    if (relative.length === 0) {
      continue;
    }
    const isDirectory = /<[a-zA-Z0-9]*:?collection\s*\/?>/.test(block);
    const length = tag(block, "getcontentlength");
    const modified = tag(block, "getlastmodified");
    const parsedModified = modified === null ? Number.NaN : Date.parse(modified);
    entries.push({
      href: relative,
      name: relative.split("/").pop() ?? relative,
      isDirectory,
      size: length === null ? 0 : Number.parseInt(length, 10) || 0,
      lastModified: Number.isNaN(parsedModified)
        ? null
        : Math.floor(parsedModified / 1000)
    });
  }
  return entries;
}

function tag(block: string, name: string): string | null {
  const match = new RegExp(
    `<[a-zA-Z0-9]*:?${name}[^>]*>([\\s\\S]*?)</[a-zA-Z0-9]*:?${name}>`
  ).exec(block);
  return match === null ? null : match[1].trim();
}

/** Content-addressed object location, sharded to keep collections small. */
export function objectPath(digest: string): string {
  if (!/^[0-9a-f]{64}$/.test(digest)) {
    throw new WebdavError(`digest must be lowercase SHA-256 hex: ${digest}`);
  }
  return `objects/${digest.slice(0, 2)}/${digest}`;
}

export function manifestPath(deviceId: string): string {
  return `manifests/${deviceId}.json`;
}

File: agent/src/relaydot/webdav.py
"""Minimal WebDAV client for the shared content-addressed object store.

Agents talk to WebDAV directly so file bytes never traverse the controller. The
controller reads the same layout to derive usage analytics, which is why the
path helpers here mirror ``apps/web/src/lib/webdav.ts`` exactly:

    objects/<first two hex>/<sha256>   immutable file content
    manifests/<device id>.json         what one device published

Only the verbs that layout needs are implemented: HEAD to test for an existing
digest, GET, PUT, MKCOL, and PROPFIND to enumerate manifests.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime
from email.utils import parsedate_to_datetime
from types import TracebackType
from typing import Any
from urllib.parse import quote, unquote, urljoin, urlsplit, urlunsplit

import httpx

from .errors import StorageError

_DIGEST = re.compile(r"^[0-9a-f]{64}$")
_RESPONSE = re.compile(r"<[a-zA-Z0-9]*:?response[\s>].*?</[a-zA-Z0-9]*:?response>", re.DOTALL)
_COLLECTION = re.compile(r"<[a-zA-Z0-9]*:?collection\s*/?>")

PROPFIND_BODY = (
    '<?xml version="1.0" encoding="utf-8"?>'
    '<d:propfind xmlns:d="DAV:"><d:prop>'
    "<d:resourcetype/><d:getcontentlength/><d:getlastmodified/>"
    "</d:prop></d:propfind>"
)


def object_path(digest: str) -> str:
    """Content-addressed location, sharded so no collection grows unbounded."""

    if not _DIGEST.match(digest):
        raise StorageError(f"digest must be lowercase SHA-256 hex: {digest}")
    return f"objects/{digest[:2]}/{digest}"


def manifest_path(device_id: str) -> str:
    if not device_id or "/" in device_id:
        raise StorageError(f"device id is not a usable path segment: {device_id!r}")
    return f"manifests/{device_id}.json"


def normalize_base_url(raw: str) -> str:
    """Return the base URL with exactly one trailing slash."""

    trimmed = raw.strip()
    if not trimmed:
        raise StorageError("WebDAV base URL must not be empty")
    parts = urlsplit(trimmed)
    if parts.scheme not in ("http", "https"):
        raise StorageError("WebDAV base URL must use http or https")
    if not parts.netloc:
        raise StorageError(f"WebDAV base URL is not absolute: {trimmed}")
    path = parts.path if parts.path.endswith("/") else f"{parts.path}/"
    return urlunsplit((parts.scheme, parts.netloc, path, "", ""))


@dataclass(frozen=True, slots=True)
class WebdavEntry:
    href: str
    name: str
    is_directory: bool
    size: int
    last_modified: int | None


def parse_propfind(xml: str, base_path: str) -> tuple[WebdavEntry, ...]:
    """Extract the four fields the agent needs from a multistatus body.

    A full XML parser is not required for href, collection flag, length, and
    mtime, and avoiding one keeps the agent free of an XML dependency whose
    entity handling would need auditing.
    """

    entries: list[WebdavEntry] = []
    for block in _RESPONSE.findall(xml):
        href = _tag(block, "href")
        if href is None:
            continue
        path = urlsplit(href).path
        relative = unquote(path)
        prefix = unquote(base_path)
        if relative.startswith(prefix):
            relative = relative[len(prefix) :]
        relative = relative.strip("/")
        if not relative:
            continue
        length = _tag(block, "getcontentlength")
        modified = _tag(block, "getlastmodified")
        entries.append(
            WebdavEntry(
                href=relative,
                name=relative.rsplit("/", 1)[-1],
                is_directory=bool(_COLLECTION.search(block)),
                size=int(length) if length and length.isdigit() else 0,
                last_modified=_http_date(modified),
            )
        )
    return tuple(entries)


def _tag(block: str, name: str) -> str | None:
    match = re.search(rf"<[a-zA-Z0-9]*:?{name}[^>]*>(.*?)</[a-zA-Z0-9]*:?{name}>", block, re.DOTALL)
    return match.group(1).strip() if match else None


def _http_date(value: str | None) -> int | None:
    if not value:
        return None
    try:
        parsed: datetime = parsedate_to_datetime(value)
    except (TypeError, ValueError):
        return None
    return int(parsed.timestamp())


class WebdavClient:
    """Basic-auth WebDAV over an injectable httpx client."""

    def __init__(
        self,
        base_url: str,
        username: str,
        password: str,
        client: httpx.Client | None = None,
        *,
        timeout: float = 60.0,
    ) -> None:
        self.base_url = normalize_base_url(base_url)
        self.base_path = urlsplit(self.base_url).path
        self._auth = (username, password)
        self._client = client or httpx.Client(timeout=timeout)
        self._owns_client = client is None

    @classmethod
    def from_config(
        cls, config: dict[str, Any], client: httpx.Client | None = None
    ) -> WebdavClient:
        """Build from the payload the controller hands an enrolled device."""

        try:
            return cls(
                str(config["base_url"]),
                str(config["username"]),
                str(config["password"]),
                client,
            )
        except KeyError as exc:
            raise StorageError(f"storage configuration is missing {exc.args[0]}") from exc

    def __enter__(self) -> WebdavClient:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        self.close()

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def url(self, path: str) -> str:
        encoded = "/".join(quote(part, safe="") for part in path.split("/") if part)
        return urljoin(self.base_url, encoded)

    def _request(
        self,
        method: str,
        path: str,
        *,
        content: bytes | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        try:
            response = self._client.request(
                method,
                self.url(path),
                auth=self._auth,
                content=content,
                headers=headers,
                follow_redirects=True,
            )
        except httpx.HTTPError as exc:
            raise StorageError(f"WebDAV {method} {path} failed: {exc}") from exc
        # Checked centrally so every verb reports a bad credential the same way
        # instead of surfacing as a confusing per-method failure.
        if response.status_code in (401, 403):
            raise StorageError("WebDAV rejected the credential")
        return response

    def exists(self, path: str) -> bool:
        """Test for a resource with a depth-0 PROPFIND rather than HEAD.

        HEAD would be cheaper, but WsgiDAV — and it is unlikely to be alone —
        answers a 404 HEAD with a ``Content-Length`` for an error page it then
        correctly does not send. That leaves a keep-alive connection expecting
        bytes that never arrive, and the *next* response on it is parsed as a
        status line and fails. Since a sync probes for many digests on one
        connection, that turns into an unusable client. PROPFIND always sends
        the body it declares, so the connection stays consistent.
        """

        response = self._request(
            "PROPFIND",
            path,
            content=PROPFIND_BODY.encode(),
            headers={"Depth": "0", "Content-Type": "application/xml"},
        )
        if response.status_code == 404:
            return False
        if response.status_code not in (200, 207):
            raise StorageError(f"WebDAV PROPFIND {path} returned {response.status_code}")
        return True

    def get_bytes(self, path: str) -> bytes | None:
        response = self._request("GET", path)
        if response.status_code == 404:
            return None
        if response.status_code >= 400:
            raise StorageError(f"WebDAV GET {path} returned {response.status_code}")
        return response.content

    def put_bytes(self, path: str, data: bytes) -> None:
        self.ensure_collection(path.rsplit("/", 1)[0] if "/" in path else "")
        response = self._request(
            "PUT",
            path,
            content=data,
            headers={"Content-Type": "application/octet-stream"},
        )
        if response.status_code >= 400:
            raise StorageError(f"WebDAV PUT {path} returned {response.status_code}")

    def mkcol(self, path: str) -> None:
        response = self._request("MKCOL", path)
        # 405 and 409-after-create races both mean "already there".
        if response.status_code >= 400 and response.status_code not in (405, 409):
            raise StorageError(f"WebDAV MKCOL {path} returned {response.status_code}")

    def ensure_collection(self, path: str) -> None:
        current = ""
        for segment in [part for part in path.split("/") if part]:
            current = segment if not current else f"{current}/{segment}"
            self.mkcol(current)

    def list(self, path: str, depth: str = "1") -> tuple[WebdavEntry, ...]:
        response = self._request(
            "PROPFIND",
            path,
            content=PROPFIND_BODY.encode(),
            headers={"Depth": depth, "Content-Type": "application/xml"},
        )
        if response.status_code == 404:
            return ()
        if response.status_code not in (200, 207):
            raise StorageError(f"WebDAV PROPFIND {path} returned {response.status_code}")
        return parse_propfind(response.text, self.base_path)
